src/daemon/global_rpc/server.rs

```diff
@@ -1,103 +1,42 @@
-use anyhow::Context;
 use async_trait::async_trait;
-use earendil_crypt::{Fingerprint, VerifyError};
-use futures_util::{stream::FuturesUnordered, StreamExt};
-use moka::sync::{Cache, CacheBuilder};
-
-use smol_timeout::TimeoutExt;
-use std::time::Duration;
-use stdcode::StdcodeSerializeExt;
-
 
-use crate::daemon::{
-    global_rpc::{transport::GlobalRpcTransport, GlobalRpcClient},
-    rendezvous::ForwardRequest,
-    DaemonContext,
-};
+use crate::daemon::rendezvous::ForwardRequest;
+use crate::daemon::{haven::HavenLocator, DaemonContext};
+use earendil_crypt::{Fingerprint, VerifyError};
 
 use super::GlobalRpcProtocol;
 
 pub struct GlobalRpcImpl {
     ctx: DaemonContext,
-
-    dht_cache: Cache<String, String>,
 }
 
 impl GlobalRpcImpl {
     pub fn new(ctx: DaemonContext) -> GlobalRpcImpl {
-        GlobalRpcImpl {
-            ctx,
-            dht_cache: CacheBuilder::default()
-                .time_to_idle(Duration::from_secs(30))
-                .build(),
-        }
-    }
-
-    fn dht_key_to_fps(&self, key: &str) -> Vec<Fingerprint> {
-        let mut all_nodes: Vec<Fingerprint> = self.ctx.relay_graph.read().all_nodes().collect();
-        all_nodes.sort_unstable_by_key(|fp| *blake3::hash(&(key, fp).stdcode()).as_bytes());
-        all_nodes
+        GlobalRpcImpl { ctx }
     }
 }
 
-const DHT_REDUNDANCY: usize = 3;
-
 #[async_trait]
 impl GlobalRpcProtocol for GlobalRpcImpl {
     async fn ping(&self, i: u64) -> u64 {
         i
     }
 
-    async fn dht_insert(&self, key: String, value: String, recurse: bool) {
-        let replicas = self.dht_key_to_fps(&key);
-
-        for replica in replicas.into_iter().take(DHT_REDUNDANCY) {
-            if replica == self.ctx.identity.public().fingerprint() {
-                log::debug!("key {key} inserting into ourselves!");
-                self.dht_cache.insert(key.clone(), value.clone());
-            } else if recurse {
-                log::debug!("key {key} inserting into remote replica {replica}");
-                let gclient = GlobalRpcClient(GlobalRpcTransport::new(self.ctx.clone(), replica));
-                match gclient
-                    .dht_insert(key.clone(), value.clone(), false)
-                    .timeout(Duration::from_secs(10))
-                    .await
-                {
-                    Some(Err(e)) => log::debug!("inserting {key} into {replica} failed: {:?}", e),
-                    None => log::debug!("inserting {key} into {replica} timed out"),
-                    _ => {}
-                }
-            }
+    async fn dht_insert(&self, key: Fingerprint, value: HavenLocator, recurse: bool) {
+        if recurse {
+            self.ctx.dht_insert(key, value).await
+        } else {
+            log::debug!("inserting key {key} locally");
```

Comment:
Here is where the validation should happen. Otherwise, bad guy can overwrite valid fp->locator mappings with bad ones. This will not fool the clients (since they should verify the signature), but it will result in disruption of service for that haven.